src/storage/src/source/postgres/replication.rs

```diff
@@ -1104,3 +1092,66 @@ async fn ensure_replication_timeline_id(
         }))
     }
 }
+
+enum SchemaValidationError {
+    Postgres(PostgresError),
+    Schema {
+        oid: u32,
+        output_index: usize,
+        error: DefiniteError,
+    },
+}
+
+fn spawn_schema_validator(
+    client: Client,
+    config: &RawSourceCreationConfig,
+    publication: String,
+    table_info: BTreeMap<u32, BTreeMap<usize, SourceOutputInfo>>,
+) -> mpsc::UnboundedReceiver<SchemaValidationError> {
+    let (tx, rx) = mpsc::unbounded_channel();
+    let source_id = config.id;
+    let config_set = Arc::clone(config.config.config_set());
+
+    mz_ore::task::spawn(|| format!("schema-validator:{}", source_id), async move {
+        while !tx.is_closed() {
+            trace!(%source_id, "validating schemas");
+
+            let upstream_info =
+                match mz_postgres_util::publication_info(&*client, &publication).await {
+                    Ok(info) => info.into_iter().map(|t| (t.oid, t)).collect(),
+                    Err(error) => {
+                        let _ = tx.send(SchemaValidationError::Postgres(error));
+                        continue;
+                    }
+                };
+
+            for (&oid, outputs) in table_info.iter() {
+                for (&output_index, output_info) in outputs {
+                    let expected_desc = &output_info.desc;
+                    let casts = &output_info.casts;
+                    if let Err(error) = verify_schema(oid, expected_desc, &upstream_info, casts) {
+                        trace!(
+                            %source_id,
+                            "schema of output index {output_index} for oid {oid} invalid",
+                        );
+                        let _ = tx.send(SchemaValidationError::Schema {
+                            oid,
+                            output_index,
+                            error,
+                        });
+                    } else {
+                        trace!(
+                            %source_id,
+                            "schema of output index {output_index} for oid {oid} valid",
+                        );
+                    }
+                }
+            }
+
+            let validation_interval = PG_SCHEMA_VALIDATION_INTERVAL.get(&config_set);
+            tokio::time::sleep(validation_interval).await;
```

Comment:
nit: this will make the validation happen every `validation_interval + schema_fetch_duration`. We can define a tokio Interval outside the loop so that we can tick every `validation_interval` reliably

Reply:
A tokio interval makes it difficult to apply updated `PG_SCHEMA_VALIDATION_INTERVAL` values at runtime, so I instead kept the `sleep` but am now accounting for how long the validation took.